torch/csrc/distributed/rpc/request_callback_impl.cpp

```diff
@@ -50,23 +50,37 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
 
       // scriptCall is only alive within this block, use reference to avoid copy
       auto& stack = scriptCall.stackRef();
+
       if (scriptCall.hasOp()) {
         scriptCall.op()->getOperation()(stack);
-      } else {
-        PythonRpcHandler::getInstance()
-            .jitCompilationUnit()
-            ->get_function(scriptCall.qualifiedName())
-            .run(stack);
+        TORCH_INTERNAL_ASSERT(
+            stack.size() == 1,
+            "Return value of a builtin operator or a "
+            "TorchScript function should be a single IValue, got a vector of "
+            "size ",
+            stack.size());
+        return wrap(
+            std::move(ScriptResp(std::move(stack.front()))).toMessage());
       }
 
-      TORCH_INTERNAL_ASSERT(
-          stack.size() == 1,
-          "Return value of a builtin operator or a "
-          "TorchScript function should be a single IValue, got a vector of "
-          "size ",
-          stack.size());
-
-      return wrap(std::move(ScriptResp(std::move(stack.front()))).toMessage());
+      auto future = PythonRpcHandler::getInstance()
+                        .jitCompilationUnit()
+                        ->get_function(scriptCall.qualifiedName())
+                        .runAsync(stack);
+      if (future->completed()) {
+        return wrap(std::move(ScriptResp(future->value())).toMessage());
+      }
+      auto responseFuture = std::make_shared<FutureMessage>();
+      future->addCallback([responseFuture, messageId, future]() {
+        try {
+          Message m = ScriptResp({future->value()}).toMessage();
+          m.setId(messageId);
+          responseFuture->markCompleted(std::move(m));
```

Comment:
can we use `wrap` for this as well?

Reply:
Not really:
  - wrap() instantiates a fresh responseFuture with an already-completed value.
  - We already have an existing responseFuture and need to complete it.

We could technically make wrap fill in an existing responseFuture, though it's more efficient not to do so.